source/s3_client.c

```diff
@@ -213,12 +213,31 @@ struct aws_s3_client *aws_s3_client_new(
         *((uint64_t *)&client_config->max_part_size) = client_config->part_size;
     }
 
-    if (client_config->tls_connection_options != NULL) {
-
+    if (client_config->tls_mode == AWS_MR_TLS_ENABLED) {
         client->tls_connection_options =
             aws_mem_calloc(client->allocator, 1, sizeof(struct aws_tls_connection_options));
+        if (client->tls_connection_options == NULL) {
+            goto on_error;
+        }
+
+        if (client_config->tls_connection_options != NULL) {
+            aws_tls_connection_options_copy(client->tls_connection_options, client_config->tls_connection_options);
+        } else {
+            struct aws_tls_ctx_options default_tls_ctx_options;
+            AWS_ZERO_STRUCT(default_tls_ctx_options);
+
+            aws_tls_ctx_options_init_default_client(&default_tls_ctx_options, allocator);
+
+            struct aws_tls_ctx *default_tls_ctx = aws_tls_client_ctx_new(allocator, &default_tls_ctx_options);
+            if (default_tls_ctx == NULL) {
+                goto on_error;
+            }
+
+            aws_tls_connection_options_init_from_ctx(client->tls_connection_options, default_tls_ctx);
 
-        aws_tls_connection_options_copy(client->tls_connection_options, client_config->tls_connection_options);
+            aws_tls_ctx_release(default_tls_ctx);
```

Comment:
How is this going to work at connection time? The aws_tls_ctx must exist for TLS to work. We could cache this on the client, but it has to live doesn't it?

Reply:
It's ref counted by the tls connection options that use it.

Reply:
OH right ok.